### PR TITLE
Change the way relays are chosen, add RelayFull message.

### PR DESCRIPTION
**Pull request by AudriusButkevicius**

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
This is compatible with the current protocol, as the client will just shit its pants about an unknown message and will try the next relay.

**Comment by AudriusButkevicius** [comment]
It compiles, haven't actually run it.

File: lib/relay/client/dynamic.go [hunk 1]
```diff
@@ -176,42 +177,46 @@ type dynamicAnnouncement struct {
...
-// by sum latency, and returns the addresses.
-func relayAddressesSortedByLatency(input []string) []string {
-	relays := make(relayList, len(input))
-	for i, relay := range input {
-		if latency, err := osutil.GetLatencyForURL(relay); err == nil {
-			relays[i] = relayWithLatency{relay, int(latency / time.Millisecond)}
-		} else {
-			relays[i] = relayWithLatency{relay, int(time.Hour / time.Millisecond)}
+// relayAddressesOrder checks the latency to each relay, rounds latency down to
+// the closest 30ms, and puts them in buckets of 50ms latency ranges. Then
```
Comment (calmh) [inline lib/relay/client/dynamic.go]: That's a weird bucketing algorithm, but I don't think that's what actually happens. The constant `30` doesn't appear in the code.

File: lib/relay/client/dynamic.go [hunk 2]
```diff
@@ -176,42 +177,46 @@ type dynamicAnnouncement struct {
...
+
+	for _, relay := range input {
+		latency, err := osutil.GetLatencyForURL(relay)
+		if err != nil {
+			latency = time.Hour
 		}
-	}
 
-	sort.Sort(relays)
+		id := int(latency/time.Millisecond) % 50
```
Comment (calmh) [inline lib/relay/client/dynamic.go]: Relays with latencies of 1 ms, 51 ms and 1001 ms will all end up in bucket 1?

File: lib/relay/client/dynamic.go [hunk 3]
```diff
@@ -176,42 +177,46 @@ type dynamicAnnouncement struct {
...
-type relayList []relayWithLatency
+	sort.Ints(ids)
 
-func (l relayList) Len() int {
-	return len(l)
-}
+	addresses := make([]string, len(input))
+
+	for _, id := range ids {
+		for _, relay := range buckets[id] {
```
Comment (calmh) [inline lib/relay/client/dynamic.go]: `addresses = append(addresses, buckets[id]...)`?

File: lib/relay/client/static.go [hunk 4]
```diff
@@ -119,6 +119,10 @@ func (c *staticClient) Serve() {
 				}
 				c.invitations <- msg
 
+			case protocol.RelayFull:
+				l.Infoln("Disconnected from relay due to it becoming full.")
```
Comment (calmh) [inline lib/relay/client/static.go]: :+1:

What happens with existing sessions here - they are retained, right?

Comment (AudriusButkevicius) [inline lib/relay/client/static.go]: You don't get dropped if you have sessions.

File: lib/relay/client/dynamic.go (hunk 2 above)
Comment (AudriusButkevicius) [inline lib/relay/client/dynamic.go]: Ha! Wrong operator!

File: lib/relay/client/dynamic.go (hunk 1 above)
Comment (AudriusButkevicius) [inline lib/relay/client/dynamic.go]: It was 30 then I changed it to 50

File: lib/relay/protocol/packets_xdr.go [hunk 5]
```diff
@@ -256,6 +256,63 @@ func (o *JoinRelayRequest) DecodeXDRFrom(xr *xdr.Reader) error {
...
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+
+struct RelayFull {
+}
+
+*/
+
+func (o RelayFull) EncodeXDR(w io.Writer) (int, error) {
```
Comment (calmh) [inline lib/relay/protocol/packets_xdr.go]: I should add a quick path to genxdr for empty structs...

**Comment by canton7** [comment]
:+1: nice

**Pull request merged**